File: python/sigil_sdk/redaction.py
"""
Secret redaction engine for Sigil content capture.

~20 high-confidence patterns hand-curated from Gitleaks
(https://github.com/gitleaks/gitleaks). Two tiers:
  - Tier 1: definite secret formats and optional email addresses
    used by both redact() and redact_lightweight()
  - Tier 2: heuristic env patterns
    used only by redact()

Add more patterns when concrete unredacted secrets are observed.
"""

from __future__ import annotations

import copy
import re
from dataclasses import dataclass

from .config import GenerationSanitizer
from .models import Generation, Message, MessageRole, Part, PartKind


@dataclass(frozen=True, slots=True)
class _SecretPattern:
    id: str
    regex: re.Pattern[str]


@dataclass(frozen=True, slots=True)
class SecretRedactionOptions:
    """
    Options for the built-in secret redaction sanitizer.

    `redact_input_messages` defaults to False to match the current opencode
    plugin behavior.

    `redact_email_addresses` defaults to True. Callers can opt out when email
    addresses should be preserved.
    """

    redact_input_messages: bool = False
    redact_email_addresses: bool = True


# --- Tier 1: High-confidence patterns (definite secret formats) ---
_TIER1_PATTERNS: tuple[_SecretPattern, ...] = (
    _SecretPattern("grafana-cloud-token", re.compile(r"\bglc_[A-Za-z0-9_-]{20,}")),
    _SecretPattern("grafana-service-account-token", re.compile(r"\bglsa_[A-Za-z0-9_-]{20,}")),
    _SecretPattern("aws-access-token", re.compile(r"\b(?:A3T[A-Z0-9]|AKIA|ASIA|ABIA|ACCA)[A-Z2-7]{16}\b")),
    _SecretPattern("github-pat", re.compile(r"\bghp_[A-Za-z0-9_]{36,}")),
    _SecretPattern("github-oauth", re.compile(r"\bgho_[A-Za-z0-9_]{36,}")),
    _SecretPattern("github-app-token", re.compile(r"\bghs_[A-Za-z0-9_]{36,}")),
    _SecretPattern("github-fine-grained-pat", re.compile(r"\bgithub_pat_[A-Za-z0-9_]{82}")),
    _SecretPattern("anthropic-api-key", re.compile(r"\bsk-ant-api03-[a-zA-Z0-9_-]{93}AA")),
    _SecretPattern("anthropic-admin-key", re.compile(r"\bsk-ant-admin01-[a-zA-Z0-9_-]{93}AA")),
    _SecretPattern("openai-api-key", re.compile(r"\bsk-[a-zA-Z0-9]{20}T3BlbkFJ[a-zA-Z0-9]{20}")),
    _SecretPattern("openai-project-key", re.compile(r"\bsk-proj-[a-zA-Z0-9_-]{40,}")),
    _SecretPattern("openai-svcacct-key", re.compile(r"\bsk-svcacct-[a-zA-Z0-9_-]{40,}")),
    _SecretPattern("gcp-api-key", re.compile(r"\bAIza[A-Za-z0-9_-]{35}")),
    _SecretPattern(
        "private-key",
        re.compile(r"-----BEGIN[A-Z ]*PRIVATE KEY-----[\s\S]*?-----END[A-Z ]*PRIVATE KEY-----"),
    ),
    _SecretPattern("connection-string", re.compile(r"(?:postgres|mysql|mongodb|redis|amqp):\/\/[^\s'\"]+@[^\s'\"]+")),
    _SecretPattern("bearer-token", re.compile(r"[Bb]earer\s+[A-Za-z0-9_.\-~+/]{20,}={0,3}")),
    _SecretPattern("slack-token", re.compile(r"\bxox[bporas]-[A-Za-z0-9-]{10,}")),
    _SecretPattern("stripe-key", re.compile(r"\b[sr]k_(?:live|test)_[A-Za-z0-9]{20,}")),
    _SecretPattern("sendgrid-api-key", re.compile(r"\bSG\.[A-Za-z0-9_-]{22}\.[A-Za-z0-9_-]{43}")),
    _SecretPattern("twilio-api-key", re.compile(r"\bSK[a-f0-9]{32}")),
    _SecretPattern("npm-token", re.compile(r"\bnpm_[A-Za-z0-9]{36}")),
    _SecretPattern("pypi-token", re.compile(r"\bpypi-[A-Za-z0-9_-]{50,}")),
)

_EMAIL_PATTERN = _SecretPattern("email", re.compile(r"\b[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}\b", re.IGNORECASE))

# --- Tier 2: Heuristic patterns (env file values) ---
_TIER2_PATTERNS: tuple[_SecretPattern, ...] = (
    _SecretPattern(
        "env-secret-value",
        re.compile(
            r"((?:PASSWORD|SECRET|TOKEN|KEY|CREDENTIAL|API_KEY|PRIVATE_KEY|ACCESS_KEY)\s*[=:]\s*)([^\s\"{}\[\],]+)",
            re.IGNORECASE,
        ),
    ),
)


class _SecretRedactor:
    """Regex-based redactor with full and lightweight modes."""

    def __init__(self, include_email_addresses: bool) -> None:
        self._include_email_addresses = include_email_addresses

    # Full redaction: tier 1 + tier 2. Use for tool call args and tool results.
    def redact(self, text: str) -> str:
        result = _apply_patterns(text, _TIER1_PATTERNS)
        if self._include_email_addresses:
            result = _apply_pattern(result, _EMAIL_PATTERN)
        return _apply_tier2_patterns(result)

    # Lightweight redaction: tier 1 only. Use for assistant text and reasoning.
    def redact_lightweight(self, text: str) -> str:
        result = _apply_patterns(text, _TIER1_PATTERNS)
        if self._include_email_addresses:
            result = _apply_pattern(result, _EMAIL_PATTERN)
        return result


def create_secret_redaction_sanitizer(
    options: SecretRedactionOptions | None = None,
) -> GenerationSanitizer:
    """Returns a reusable generation sanitizer that redacts known secret formats."""

    resolved = options or SecretRedactionOptions()
    redactor = _SecretRedactor(include_email_addresses=resolved.redact_email_addresses)

    def _sanitize(generation: Generation) -> Generation:
        for message in generation.input:
            mode = "full" if message.role == MessageRole.USER and resolved.redact_input_messages else "none"
            _sanitize_message(message, redactor, mode)

        for message in generation.output:
            if message.role == MessageRole.ASSISTANT:
                mode = "light"
            elif message.role == MessageRole.TOOL:
                mode = "full"
            else:
                mode = "none"
            _sanitize_message(message, redactor, mode)

        return generation

    return _sanitize


def _sanitize_message(message: Message, redactor: _SecretRedactor, default_text_mode: str) -> None:
    for part in message.parts:
        _sanitize_part(part, redactor, default_text_mode)


def _sanitize_part(part: Part, redactor: _SecretRedactor, default_text_mode: str) -> None:
    if default_text_mode == "none":
        return
    if part.kind == PartKind.TEXT:
        part.text = _redact_string(part.text, redactor, default_text_mode)
        return
    if part.kind == PartKind.THINKING:
        part.thinking = redactor.redact_lightweight(part.thinking)
        return
    if part.kind == PartKind.TOOL_CALL and part.tool_call is not None:
        if len(part.tool_call.input_json) > 0:
            part.tool_call.input_json = redactor.redact(part.tool_call.input_json.decode("utf-8")).encode("utf-8")
        return
    if part.kind == PartKind.TOOL_RESULT and part.tool_result is not None:
        part.tool_result.content = redactor.redact(part.tool_result.content)
        if len(part.tool_result.content_json) > 0:
            part.tool_result.content_json = redactor.redact(part.tool_result.content_json.decode("utf-8")).encode(
                "utf-8"
            )


def _redact_string(value: str, redactor: _SecretRedactor, mode: str) -> str:
    if mode == "full":
        return redactor.redact(value)
    if mode == "light":
        return redactor.redact_lightweight(value)
    return value


def _apply_patterns(text: str, patterns: tuple[_SecretPattern, ...]) -> str:
    result = text
    for pattern in patterns:
        result = _apply_pattern(result, pattern)
    return result


def _apply_pattern(text: str, pattern: _SecretPattern) -> str:
    return pattern.regex.sub(f"[REDACTED:{pattern.id}]", text)


def _apply_tier2_patterns(text: str) -> str:
    result = text
    for pattern in _TIER2_PATTERNS:
        result = pattern.regex.sub(rf"\1[REDACTED:{pattern.id}]", result)
    return result

File: js/src/redaction.ts
import type { GenerationSanitizer, Message, MessagePart } from './types.js';
import { cloneGeneration } from './utils.js';

/**
 * Secret redaction engine for Sigil content capture.
 *
 * ~20 high-confidence patterns hand-curated from Gitleaks
 * (https://github.com/gitleaks/gitleaks). Two tiers:
 *   - Tier 1: definite secret formats — used by both redact() and redactLightweight()
 *   - Tier 2: heuristic env patterns — used only by redact()
 *
 * Add more patterns when concrete unredacted secrets are observed.
 */

interface SecretPattern {
  id: string;
  regex: RegExp;
}

export interface SecretRedactionOptions {
  /**
   * Redact user input messages in addition to assistant/tool content.
   * Defaults to `false` to match the current opencode plugin behavior.
   */
  redactInputMessages?: boolean;
  /**
   * Redact generic email addresses.
   * Defaults to `true`. Set to `false` to opt out when company policy allows
   * email-like content.
   */
  redactEmailAddresses?: boolean;
}

// --- Tier 1: High-confidence patterns (definite secret formats) ---
const tier1Patterns: SecretPattern[] = [
  { id: 'grafana-cloud-token', regex: /\bglc_[A-Za-z0-9_-]{20,}/g },
  { id: 'grafana-service-account-token', regex: /\bglsa_[A-Za-z0-9_-]{20,}/g },
  { id: 'aws-access-token', regex: /\b(?:A3T[A-Z0-9]|AKIA|ASIA|ABIA|ACCA)[A-Z2-7]{16}\b/g },
  { id: 'github-pat', regex: /\bghp_[A-Za-z0-9_]{36,}/g },
  { id: 'github-oauth', regex: /\bgho_[A-Za-z0-9_]{36,}/g },
  { id: 'github-app-token', regex: /\bghs_[A-Za-z0-9_]{36,}/g },
  { id: 'github-fine-grained-pat', regex: /\bgithub_pat_[A-Za-z0-9_]{82}/g },
  { id: 'anthropic-api-key', regex: /\bsk-ant-api03-[a-zA-Z0-9_-]{93}AA/g },
  { id: 'anthropic-admin-key', regex: /\bsk-ant-admin01-[a-zA-Z0-9_-]{93}AA/g },
  { id: 'openai-api-key', regex: /\bsk-[a-zA-Z0-9]{20}T3BlbkFJ[a-zA-Z0-9]{20}/g },
  { id: 'openai-project-key', regex: /\bsk-proj-[a-zA-Z0-9_-]{40,}/g },
  { id: 'openai-svcacct-key', regex: /\bsk-svcacct-[a-zA-Z0-9_-]{40,}/g },
  { id: 'gcp-api-key', regex: /\bAIza[A-Za-z0-9_-]{35}/g },
  { id: 'private-key', regex: /-----BEGIN[A-Z ]*PRIVATE KEY-----[\s\S]*?-----END[A-Z ]*PRIVATE KEY-----/g },
  { id: 'connection-string', regex: /(?:postgres|mysql|mongodb|redis|amqp):\/\/[^\s'"]+@[^\s'"]+/g },
  { id: 'bearer-token', regex: /[Bb]earer\s+[A-Za-z0-9_.\-~+/]{20,}={0,3}/g },
  { id: 'slack-token', regex: /\bxox[bporas]-[A-Za-z0-9-]{10,}/g },
  { id: 'stripe-key', regex: /\b[sr]k_(?:live|test)_[A-Za-z0-9]{20,}/g },
  { id: 'sendgrid-api-key', regex: /\bSG\.[A-Za-z0-9_-]{22}\.[A-Za-z0-9_-]{43}/g },
  { id: 'twilio-api-key', regex: /\bSK[a-f0-9]{32}/g },
  { id: 'npm-token', regex: /\bnpm_[A-Za-z0-9]{36}/g },
  { id: 'pypi-token', regex: /\bpypi-[A-Za-z0-9_-]{50,}/g },
];

const emailPattern: SecretPattern = {
  id: 'email',
  regex: /\b[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}\b/gi,
};

// --- Tier 2: Heuristic patterns (env file values) ---
const tier2Patterns: SecretPattern[] = [
  {
    id: 'env-secret-value',
    regex: /((?:PASSWORD|SECRET|TOKEN|KEY|CREDENTIAL|API_KEY|PRIVATE_KEY|ACCESS_KEY)\s*[=:]\s*)(\S+)/gi,
  },
];

class SecretRedactor {
  private readonly includeEmailAddresses: boolean;

  constructor(includeEmailAddresses: boolean) {
    this.includeEmailAddresses = includeEmailAddresses;
  }

  /** Full redaction: tier 1 + tier 2. Use for tool call args and tool results. */
  redact(text: string): string {
    let result = applyPatterns(text, tier1Patterns);
    if (this.includeEmailAddresses) {
      result = applyPattern(result, emailPattern);
    }
    return applyTier2Patterns(result, tier2Patterns);
  }

  /** Lightweight redaction: tier 1 only. Use for assistant text and reasoning. */
  redactLightweight(text: string): string {
    let result = applyPatterns(text, tier1Patterns);
    if (this.includeEmailAddresses) {
      result = applyPattern(result, emailPattern);
    }
    return result;
  }
}

export function createSecretRedactionSanitizer(options: SecretRedactionOptions = {}): GenerationSanitizer {
  const redactor = new SecretRedactor(options.redactEmailAddresses ?? true);
  const redactInputMessages = options.redactInputMessages ?? false;

  return (generation) => {
    const sanitized = cloneGeneration(generation);

    if (sanitized.systemPrompt !== undefined) {
      sanitized.systemPrompt = redactor.redactLightweight(sanitized.systemPrompt);
    }
    if (sanitized.conversationTitle !== undefined) {
      sanitized.conversationTitle = redactor.redactLightweight(sanitized.conversationTitle);
    }
    if (sanitized.callError !== undefined) {
      sanitized.callError = redactor.redactLightweight(sanitized.callError);
    }

    for (const message of sanitized.input ?? []) {
      sanitizeMessage(message, redactor, message.role === 'user' && redactInputMessages ? 'full' : 'none');
    }
    for (const message of sanitized.output ?? []) {
      sanitizeMessage(
        message,
        redactor,
        message.role === 'assistant' ? 'light' : message.role === 'tool' ? 'full' : 'none',
      );
    }

    return sanitized;
  };
}

function sanitizeMessage(message: Message, redactor: SecretRedactor, defaultTextMode: 'none' | 'light' | 'full'): void {
  if (typeof message.content === 'string') {
    message.content = redactString(message.content, redactor, defaultTextMode);
  }
  for (const part of message.parts ?? []) {
    sanitizePart(part, redactor, defaultTextMode);
  }
}

function sanitizePart(part: MessagePart, redactor: SecretRedactor, defaultTextMode: 'none' | 'light' | 'full'): void {
  switch (part.type) {
    case 'text':
      part.text = redactString(part.text, redactor, defaultTextMode);
      break;
    case 'thinking':
      if (defaultTextMode !== 'none') {
        part.thinking = redactor.redactLightweight(part.thinking);
      }
      break;
    case 'tool_call':
      if (defaultTextMode !== 'none' && typeof part.toolCall.inputJSON === 'string') {
        part.toolCall.inputJSON = redactor.redact(part.toolCall.inputJSON);
      }
      break;
    case 'tool_result':
      if (defaultTextMode !== 'none') {
        if (typeof part.toolResult.content === 'string') {
          part.toolResult.content = redactor.redact(part.toolResult.content);
        }
        if (typeof part.toolResult.contentJSON === 'string') {
          part.toolResult.contentJSON = redactor.redact(part.toolResult.contentJSON);
        }
      }
      break;
  }
}

function redactString(value: string, redactor: SecretRedactor, mode: 'none' | 'light' | 'full'): string {
  switch (mode) {
    case 'full':
      return redactor.redact(value);
    case 'light':
      return redactor.redactLightweight(value);
    default:
      return value;
  }
}

function applyPatterns(text: string, patterns: SecretPattern[]): string {
  let result = text;
  for (const pattern of patterns) {
    result = applyPattern(result, pattern);
  }
  return result;
}

function applyPattern(text: string, pattern: SecretPattern): string {
  pattern.regex.lastIndex = 0;
  return text.replace(pattern.regex, `[REDACTED:${pattern.id}]`);
}

function applyTier2Patterns(text: string, patterns: SecretPattern[]): string {
  let result = text;
  for (const pattern of patterns) {
    pattern.regex.lastIndex = 0;
    result = result.replace(pattern.regex, `$1[REDACTED:${pattern.id}]`);
  }
  return result;
}
